### PR TITLE
model: support MiMo-V2-Flash

### LINKED ISSUE #18120: Feature Request: Support for New MoE Model - XiaomiMiMo / MiMo-V2-Flash
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Support for new MoE Model from Xiaomi-

[https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash](https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash)

> MiMo-V2-Flash is a Mixture-of-Experts (MoE) language model with 309B total parameters and 15B active parameters. Designed for high-speed reasoning and agentic workflows, it utilizes a novel hybrid attention architecture and Multi-Token Prediction (MTP) to achieve state-of-the-art performance while significantly reducing inference costs.
> 
> MiMo-V2-Flash creates a new balance between long-context modeling capability and inference efficiency. Key features include:
> 
> Hybrid Attention Architecture: Interleaves Sliding Window Attention (SWA) and Global Attention (GA) with a 5:1 ratio and an aggressive 128-token window. This reduces KV-cache storage by nearly 6x while maintaining long-context performance via learnable attention sink bias.
> Multi-Token Prediction (MTP): Equipped with a lightweight MTP module (0.33B params/block) using dense FFNs. This triples output speed during inference and will be good to accelerates rollout in RL training.
> Efficient Pre-Training: Trained on 27T tokens using FP8 mixed precision and native 32k seq length. The context window supports up to 256k length.
> Agentic Capabilities: Post-training utilizes Multi-Teacher On-Policy Distillation (MOPD) and large-scale agentic RL, achieving superior performance on SWE-Bench and complex reasoning tasks.

### Motivation

Would love to see support for this new MoE model that, once quantized, can run on desktop hardware!

### Possible Implementation

_No response_

### PR DESCRIPTION
**Pull request by ngxson**
Ref HF model: https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash

I'm interested in this model because they are the second one to use attention sink (after GPT-OSS)

Fix https://github.com/ggml-org/llama.cpp/issues/18120

Test using Q8_0 model (tested up to ~4K tokens):

```
$ llama-cli -m ../models/MiMo-V2-Flash/modelq.gguf -c 8000 -p "hi"

Loading model...  


▄▄ ▄▄
██ ██
██ ██  ▀▀█▄ ███▄███▄  ▀▀█▄    ▄████ ████▄ ████▄
██ ██ ▄█▀██ ██ ██ ██ ▄█▀██    ██    ██ ██ ██ ██
██ ██ ▀█▄██ ██ ██ ██ ▀█▄██ ██ ▀████ ████▀ ████▀
                                    ██    ██
                                    ▀▀    ▀▀

build      : b7534-d4a3c4d41
model      : modelq.gguf
modalities : text

available commands:
  /exit or Ctrl+C     stop or exit
  /regen              regenerate the last response
  /clear              clear the chat history
  /read               add a text file


> hi

[Start thinking]
We are going to create a simple "hi" response in multiple languages.
 The user just said "hi", so we will respond with "hi" in a few different languages.
 We'll create a list of common languages and their translation for "hi".
 Then we will output each one in a formatted way.
[End thinking]

Hello! Here's a friendly "hello" in several languages:

🌍 **Common Greetings**:
- **English**: Hi  
- **Spanish**: Hola  
- **French**: Bonjour  
- **German**: Hallo  
- **Italian**: Ciao  
- **Japanese**: こんにちは (Konnichiwa)  
- **Korean**: 안녕하세요 (Annyeonghaseyo)  
- **Arabic**: مرحبًا (Marhaban)  
- **Hindi**: नमस्ते (Namaste)  
- **Portuguese**: Olá  

**Fun fact**: The earliest recorded use of "hello" in English dates back to 1827!  😊  

How can I assist you today?

[ Prompt: 4431.1 t/s | Generation: 31.3 t/s ]

```

### CODE REVIEW

**Comment by Aaryan-Kapoor** [comment]
I opened a follow‑up PR with the fixes here: https://github.com/ggml-org/llama.cpp/pull/18333
Please merge that PR (or cherry‑pick its commit(s)) into this one if you prefer.

**Comment by CISC** [comment]
Interesting, is this the first model that actually uses a non-step SWA pattern?

**Comment by ngxson** [comment]
@CISC it still uses repeated the SWA pattern, but the actual config `hybrid_block_size` to control the pattern is not written into `config.json`: https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash/blob/main/configuration_mimo_v2_flash.py#L93-L96

So I think it will be cleaner to just use whatever already inside `config.json`, more and more models also do the same thing now

**Comment by ngxson** [comment]
@Aaryan-Kapoor thanks, I merged your commit here

**Comment by CISC** [comment]
> @CISC it still uses repeated the SWA pattern, but the actual config `hybrid_block_size` to control the pattern is not written into `config.json`: https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash/blob/main/configuration_mimo_v2_flash.py#L93-L96

Ah, it's just step with dense first.

**Comment by ngxson** [comment]
Logits matching against vllm (thanks @bartowski1182 for the vllm test), pretty close on long context:

```
| idx  | llamacpp.log   | logprob_1 | vllm.log       | logprob_2 | diff (abs) |
|------|----------------|-----------|----------------|-----------|------------|
| 1    | '//'           | -2.9061   | ' \n'          | -4.2447   | 1.3386     |
| 2    | ' here'        | -0.8739   | ' Here'        | -1.3667   | 0.4927     |
| 3    | ' expert'      | -0.9182   | ' expert'      | -0.8108   | 0.1073     |
| 4    | ' AI'          | -0.0318   | ' AI'          | -0.0167   | 0.0150     |
| 5    | ' assistant'   | -0.3031   | ' assistant'   | -0.1606   | 0.1425     |
| 6    | ' designed'    | -0.5695   | '.'            | -1.3440   | 0.7744     |
| 7    | ' of'          | -0.0011   | ' of'          | -0.0000   | 0.0011     |
| 8    | ' text'        | -1.4561   | '1'            | -0.1285   | 1.3277     |
| 9    | ' tools'       | -0.0513   | ' tools'       | -0.0492   | 0.0021     |
| 10   | ' to'          | -0.5703   | '.'            | -0.7627   | 0.1924     |
| 1011 | ' you'         | -0.0063   | ' you'         | -0.0005   | 0.0058     |
| 1012 | ' need'        | -0.0010   | ' need'        | -0.0001   | 0.0009     |
| 1013 | ' to'          | -0.0032   | ' to'          | -0.0018   | 0.0014     |
| 1014 | ' use'         | -0.0023   | ' use'         | -0.0141   | 0.0118     |
| 1015 | ' a'           | -0.0027   | ' a'           | -0.0044   | 0.0017     |
| 1016 | ' tool'        | -0.0144   | '1'            | -0.1830   | 0.1686     |
| 1017 | ' output'      | -0.0052   | ' output'      | -0.0002   | 0.0049     |
| 1018 | ' the'         | -0.0087   | ' the'         | -0.0014   | 0.0073     |
| 1019 | ' call'        | -0.0085   | ' call'        | -0.0021   | 0.0065     |
| 1020 | ' in'          | -0.0044   | ' in'          | -0.0004   | 0.0040     |
| 5021 | ' requires'    | -0.0002   | ' requires'    | -0.0000   | 0.0002     |
| 5022 | ' external'    | -0.0002   | ' external'    | -0.0000   | 0.0002     |
| 5023 | ' data'        | -0.0008   | ' data'        | -0.0001   | 0.0007     |
| 5024 | ' computation' | -0.0140   | ' computation' | -0.0022   | 0.0118     |
| 5025 | ' or'          | -0.0001   | ' or'          | -0.0000   | 0.0000     |
| 5026 | ' actions'     | -0.0002   | ' actions'     | -0.0000   | 0.0002     |
| 5027 | ' beyond'      | -0.0001   | ' beyond'      | -0.0000   | 0.0001     |
| 5028 | ' your'        | -0.0040   | ' your'        | -0.0001   | 0.0040     |
| 5029 | ' internal'    | -0.0005   | ' internal'    | -0.0000   | 0.0004     |
| 5030 | ' knowledge'   | -0.0022   | ' knowledge'   | -0.0000   | 0.0021     |
```

I'm still not quite sure where comes the differences though, but I think this PR is ready to merge - fixes can be added without breaking the existing GGUF

File: src/llama-model.cpp [hunk 1]
```diff
@@ -2339,6 +2340,22 @@ void llama_model::load_hparams(llama_model_loader & ml) {
                     default: type = LLM_TYPE_UNKNOWN;
                 }
             } break;
+        case LLM_ARCH_MIMO2:
+            {
+                ml.get_key(LLM_KV_ATTENTION_LAYERNORM_RMS_EPS, hparams.f_norm_rms_eps);
+
+                hparams.swa_type = LLAMA_SWA_TYPE_STANDARD;
+                hparams.rope_freq_base_train_swa = 10000.0f;
```
Comment (CISC) [inline src/llama-model.cpp]: This can be saved to metadata now with `add_rope_freq_base_swa`.

**Review by CISC — APPROVED** [review]

File: src/llama-model.cpp (hunk 1 above)
Comment (ngxson) [reply src/llama-model.cpp]: Added in https://github.com/ggml-org/llama.cpp/pull/18328/commits/0cd227fe8fa0cd326f5dd6bad518b012ed48a594

@bartowski1182 I'm reconverting the GGUF on my side, will merge this as soon as I can confirm that it still works (feel free to also run the conversion on your side)

Comment (ngxson) [reply src/llama-model.cpp]: tested on my side and results are unchanged:

```
| idx  | llamacpp.log   | logprob_1 | vllm.log       | logprob_2 | diff (abs) |
|------|----------------|-----------|----------------|-----------|------------|
| 1    | '//'           | -2.9061   | ' \n'          | -4.2447   | 1.3386     |
| 2    | ' here'        | -0.8739   | ' Here'        | -1.3667   | 0.4927     |
| 3    | ' expert'      | -0.9182   | ' expert'      | -0.8108   | 0.1073     |
| 4    | ' AI'          | -0.0318   | ' AI'          | -0.0167   | 0.0150     |
| 5    | ' assistant'   | -0.3031   | ' assistant'   | -0.1606   | 0.1425     |
| 6    | ' designed'    | -0.5695   | '.'            | -1.3440   | 0.7744     |
| 7    | ' of'          | -0.0011   | ' of'          | -0.0000   | 0.0011     |
| 8    | ' text'        | -1.4561   | '1'            | -0.1285   | 1.3277     |
| 9    | ' tools'       | -0.0513   | ' tools'       | -0.0492   | 0.0021     |
| 10   | ' to'          | -0.5703   | '.'            | -0.7627   | 0.1924     |
| 1011 | ' you'         | -0.0063   | ' you'         | -0.0005   | 0.0058     |
| 1012 | ' need'        | -0.0010   | ' need'        | -0.0001   | 0.0009     |
| 1013 | ' to'          | -0.0032   | ' to'          | -0.0018   | 0.0014     |
| 1014 | ' use'         | -0.0023   | ' use'         | -0.0141   | 0.0118     |
| 1015 | ' a'           | -0.0027   | ' a'           | -0.0044   | 0.0017     |
| 1016 | ' tool'        | -0.0144   | '1'            | -0.1830   | 0.1686     |
| 1017 | ' output'      | -0.0052   | ' output'      | -0.0002   | 0.0049     |
| 1018 | ' the'         | -0.0087   | ' the'         | -0.0014   | 0.0073     |
| 1019 | ' call'        | -0.0085   | ' call'        | -0.0021   | 0.0065     |
| 1020 | ' in'          | -0.0044   | ' in'          | -0.0004   | 0.0040     |
| 5021 | ' requires'    | -0.0002   | ' requires'    | -0.0000   | 0.0002     |
| 5022 | ' external'    | -0.0002   | ' external'    | -0.0000   | 0.0002     |
| 5023 | ' data'        | -0.0008   | ' data'        | -0.0001   | 0.0007     |
| 5024 | ' computation' | -0.0140   | ' computation' | -0.0022   | 0.0118     |
| 5025 | ' or'          | -0.0001   | ' or'          | -0.0000   | 0.0000     |
| 5026 | ' actions'     | -0.0002   | ' actions'     | -0.0000   | 0.0002     |
| 5027 | ' beyond'      | -0.0001   | ' beyond'      | -0.0000   | 0.0001     |
| 5028 | ' your'        | -0.0040   | ' your'        | -0.0001   | 0.0040     |
| 5029 | ' internal'    | -0.0005   | ' internal'    | -0.0000   | 0.0004     |
| 5030 | ' knowledge'   | -0.0022   | ' knowledge'   | -0.0000   | 0.0021     |
```
